sympy/solvers/diophantine/diophantine.py

```diff
@@ -51,6 +52,86 @@
     "univariate"}
 
 
+class DiophantineSolutionSet(set):
+    """
+    Container for a set of solutions to a particular diophantine equation.
+
+    The base representation is a set of tuples representing each of the solutions.
+
+    Parameters
+    ==========
+
+    symbols : list
+        List of free symbols in the original equation.
+    parameters: list (optional)
+        List of parameters to be used in the solution
+
+    Examples
+    ========
+
+    Adding solutions:
+
+        >>> from sympy.solvers.diophantine.diophantine import DiophantineSolutionSet
+        >>> from sympy.abc import x, y, t, u
+        >>> s1 = DiophantineSolutionSet([x, y], [t, u])
+        >>> s1
+        set()
+        >>> s1.add(2, 3)
+        >>> s1.add(-1, u)
+        >>> s1
+        {(-1, u), (2, 3)}
+        >>> s2 = DiophantineSolutionSet([x, y], [t, u])
+        >>> s2.add(3, 4)
+        >>> s1.update(s2)
+        >>> s1
+        {(-1, u), (2, 3), (3, 4)}
+
+    Conversion of solutions into dicts:
+
+        >>> list(s1.dict_iterator())
+        [{x: -1, y: u}, {x: 2, y: 3}, {x: 3, y: 4}]
+
+    Substituting values:
+
+        >>> s3 = DiophantineSolutionSet([x, y], [t, u])
+        >>> s3.add(t**2, t + u)
+        >>> s3
+        {(t**2, t + u)}
+        >>> s3.substitute_parameters(2, 3)
+        {(4, 5)}
+        >>> s3.substitute_parameters(-1)
+        {(1, u - 1)}
+    """
+
+    def __init__(self, symbols_list, parameters=None):
+        super().__init__()
+        self.symbols = symbols_list
+
+        if parameters is None:
+            self.parameters = symbols('%s1:%i' % ('t', len(self.symbols) + 1), integer=True)
+        else:
+            self.parameters = parameters
+
+    def add(self, *solution):
+        if len(solution) != len(self.symbols):
+            raise ValueError("Solution is of incorrect length")
+        super(DiophantineSolutionSet, self).add(Tuple(*solution))
+
+    def dict_iterator(self):
+        for solution in sorted(self, key=lambda s: tuple(s)):
+            yield dict(zip(self.symbols, solution))
+
+    def substitute_parameters(self, *values):
```

Comment:
Maybe there is a better name for this. Or what about simply giving it a `__call__` method so `DiophantineSolutionSet(..)(1)` gives the set with the first parameter set to 1.

Thinking more about this suggests that the API should be `(n, m, params)` where `n` is the solution length and m is the number of parameters to be used. And substitution could accept a dict mapping parameter number to parameter value:
```
D = DiophantineSoutionSet(2, 2)
D.add(x, y**2)
D.subs(1)
{(1, y**2)}
D(1)
{1, y**2}
D.subs({1:2})
{(x, 4)}
```
Perhaps the set would raise an error of the number of parameters added ever exceeded the specified parameter length, e.g. D(2,1).add(x,y) would raise an error because there are 2 free symbols but there are only supposed to be 1.

Reply:
What are you trying to fix by adding this new object? I was just thinking that this "parameterized set" could be made by a user for convenience in seeing solutions, e.g.
```
>>> sol = diophantine(x**2+y**2-z**2)
>>> sol
{(2*p*q, p**2 - q**2, p**2 + q**2)}
>>> t = Tuple(*ordered(sol))
>>> T=Lambda(tuple(ordered(t.free_symbols)), t)
>>> T
Lambda((p, q), ((2*p*q, p**2 - q**2, p**2 + q**2),))
>>> T(1,2)
((4, -3, 5),
```

Reply:
I don't know diophantine so well but I imagine it is similar to the SingleODEProblem class in #18403. The problem is that internally in dsolve there are lots of places where you don't know if something is an Eq, Expr, list of Eq etc. Making a new class makes it clear what type of object something is and forces all uses of it to implement a proper conversion at the right place.

I wonder if something like this parametrised solution set is needed more generally in solvers. It could be returned by e.g. linssolve.

Reply:
Having a parameterized set would be like PurePoly in that it could be made canonical in terms of parameters used, so `solve(z-y,z)->[y] ->{z: ParameterSet(t,Tuple(t))}` and `solve(a-b,a)->[b]->{a: ParameterSet(t,Tuple(t))}`.

Reply:
I guess a general parametrised solution set should be more or less like an ImageSet in that you have base sets as well as the solution expression and symbols (or maybe #18497)